context.go

```diff
@@ -54,12 +54,26 @@ func NewContext(handle *Handle) *Context {
 	return &Context{handle: handle, cContext: cContext}
 }
 
-// Run encodes the given addressesToData values and runs them against the WAF rules within the given
-// timeout value. It returns the matches as a JSON string (usually opaquely used) along with the corresponding
-// actions in any. In case of an error, matches and actions can still be returned, for instance in the case of a
-// timeout error. Errors can be tested against the RunError type.
-func (context *Context) Run(addressesToData map[string]any, timeout time.Duration) (res Result, err error) {
-	if len(addressesToData) == 0 {
+// RunAddressData provides address data to the Context.Run method.
+type RunAddressData struct {
+	// persistent address data is scoped to the lifetime of a given Context, and subsquent calls to Context.Run with the
+	// same address name will be silently ignored.
+	persistent map[string]any
+	// ephemeral address data is scoped to a given Context.Run call and is not persisted across calls. This is used for
+	// protocols such as gRPC client/server streaming or GraphQL, where a single request can incur multiple subrequests.
+	ephemeral map[string]any
+}
+
+func (d RunAddressData) isEmpty() bool {
+	return len(d.persistent) == 0 && len(d.ephemeral) == 0
+}
+
+// Run encodes the given persistentAddressToData and ephemeralAddressToData values and runs them against the WAF rules
+// within the given timeout value. It returns the matches as a JSON string (usually opaquely used) along with the
+// corresponding actions in any. In case of an error, matches and actions can still be returned, for instance in the
+// case of a timeout error. Errors can be tested against the RunError type.
+func (context *Context) Run(addressData RunAddressData, timeout time.Duration) (res Result, err error) {
+	if addressData.isEmpty() {
 		return
 	}
 
```

Comment:
Let's add something to make sure both `addressData.persistent` and `addressData.ephemeral` cannot be nil at the same time

Reply:
There's an existing test that asserts correct behavior with `nil` addresses, and I think we should preserve the same semantics (which is, it's okay and it's a no-op).

Reply:
I thought this was not coveree. Thanks for the explanation